### PR TITLE
Added Snipe card

### PR DESCRIPTION
**Pull request by aoaaceai**

### CODE REVIEW

File: src/main/java/eod/Character.java [hunk 1]
```diff
@@ -3,17 +3,26 @@
...
     public boolean isTargeted;
     public Point position;
     public boolean isAttacked = false;
+    private int max_hp = 30;
+    private int hp;
 
     public Character(Player player, int x, int y) {
         this.player = player;
         position = new Point(x, y);
     }
```
Comment (secminhr) [inline src/main/java/eod/Character.java]: hp will be null if using this constructor

File: src/main/java/eod/Character.java [hunk 2]
```diff
@@ -24,6 +33,26 @@ public void move(int steps) {
         }
     }
 
+    public void heal(int gain) {
+        if(hp+gain >= max_hp) {
+            hp = max_hp;
+        }
+        else {
```
Comment (secminhr) [inline src/main/java/eod/Character.java]: please use 
```} else {```

File: src/main/java/eod/Character.java [hunk 3]
```diff
@@ -59,4 +88,14 @@ private boolean addPointIfEmpty(ArrayList<Point> points, int x, int y, Gameboard
...
+
+    private void die() {
+        player.loseCharacter(this);
+        teardown();
+    }
+
+    @Override
+    public void teardown() {
+        // Nothing to do
+    }
```
Comment (secminhr) [inline src/main/java/eod/Character.java]: release player

File: src/main/java/eod/Gameboard.java [hunk 4]
```diff
@@ -37,6 +37,16 @@ public void moveElement(Point from, Point to) throws MoveInvalidException, Array
         board[from.x][from.y] = null;
     }
 
+    public void removeObject(int x, int y) throws NullPointerException, ArrayIndexOutOfBoundsException{
+        if(x<0 || x >= boardSize || y < 0 || y >= boardSize) {
+            throw new ArrayIndexOutOfBoundsException("Trying to remove a character at ("+x+", "+y+").");
+        }
+        if(board[x][y] == null) {
+            throw new NullPointerException("There's nothing at ("+x+", "+y+") on the board, cannot remove.");
```
Comment (secminhr) [inline src/main/java/eod/Gameboard.java]: should use IllegalArgumentException

File: src/main/java/eod/card/concrete/attack/Snipe.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public String getName() {
+        return null;
+    }
+
+    @Override
+    public CardParty getParty() {
+        return null;
+    }
```
Comment (secminhr) [inline src/main/java/eod/card/concrete/attack/Snipe.java]: finish ```getName``` and ```getParty```

File: src/main/java/eod/card/concrete/attack/Snipe.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public Card copy() {
+        return new Snipe(player);
+    }
+
+    @Override
+    public void attack() {
+        Accessing characters = Character(player.getBoard());
+        RequestAttack(player, 8)
+                .from(characters.which(OwnedBy(player)).which(BelongsTo(Sniper.class)).get())
```
Comment (secminhr) [inline src/main/java/eod/card/concrete/attack/Snipe.java]: is or being

File: src/main/java/eod/effect/Heal.java [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
     }
 
-    public void on(WarObject[] targets) {
-
+    public void from(Character[] characters) {
+        Character selected = askToSelect(player, characters);
+        selected.heal(hp);
     }
 
     @Override
```
Comment (secminhr) [inline src/main/java/eod/effect/Heal.java:21]: release player

File: src/main/java/eod/effect/Move.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    public void from(Character[] characters) {
+        Character selected = askToSelect(player, characters);
+        selected.move(step);
+    }
+
+    @Override
+    public void teardown() {
+        //intentionally left blank
+    }
```
Comment (secminhr) [inline src/main/java/eod/effect/Move.java]: release player

File: src/main/java/eod/listener/Listener.java [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package eod.listener;
```
Comment (secminhr) [inline src/main/java/eod/listener/Listener.java]: delete

File: src/main/java/eod/Game.java [hunk 10]
```diff
@@ -79,12 +79,25 @@ public void start() {
...
+
     private boolean handIsInvalid(Player player) {
         return !player.checkInHand(ActionCard.class);
     }
 
     private void gameLoop() throws GameLosingException {
+        //TODO: when a round starts, tells the rival player to add listeners
+    }
 
+    public void targetedTrigger(Player from) {
```
Comment (secminhr) [inline src/main/java/eod/Game.java]: trigerTargetedListener

**Review by secminhr — COMMENTED** [review]
Fix commented issues

**Comment by aoaaceai** [comment]
Fixed, I think.

File: src/main/java/eod/Character.java (hunk 2 above)
Comment (secminhr) [reply src/main/java/eod/Character.java]: Not fixed

**Comment by secminhr** [comment]
Please do not add new things. It's time to take a break and review our design.

**Review by secminhr — APPROVED** [review]

**Pull request merged**